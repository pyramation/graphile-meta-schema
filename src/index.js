import { makeExtendSchemaPlugin, gql } from 'graphile-utils';
const GIS_TYPES = [
  'Geometry',
  'Point',
  'LineString',
  'Polygon',
  'MultiPoint',
  'MultiLineString',
  'MultiPolygon',
  'GeometryCollection'
];
export const PgMetaschemaPlugin = makeExtendSchemaPlugin(
  (build, schemaOptions) => {
    /** @type {import('graphile-build-pg').PgIntrospectionResultsByKind} */
    const introspection = build.pgIntrospectionResultsByKind;
    const inflection = build.inflection;

    /** @type {string[]} */
    const schemas = schemaOptions.pgSchemas;

    return {
      typeDefs: gql`
        type MetaschemaType {
          name: String!
          modifier: Int
          typmod: JSON
          isArray: Boolean!
        }
        type MetaschemaField {
          name: String!
          type: MetaschemaType!
        }
        type MetaschemaTableInflection {
          # https://github.com/graphile/graphile-engine/blob/v4/packages/graphile-build-pg/src/plugins/PgBasicsPlugin.js
          allRows: String!
          allRowsSimple: String!
          tableFieldName: String!
          tableType: String!
          createPayloadType: String!
          updatePayloadType: String!
          deletePayloadType: String!
          deleteByPrimaryKey: String
          updateByPrimaryKey: String
          createField: String!
        }
        type MetaschemaTable {
          name: String!
          inflection: MetaschemaTableInflection!
          fields: [MetaschemaField]
          constraints: [MetaschemaConstraint]
          foreignKeyConstraints: [MetaschemaForeignKeyConstraint]
          primaryKeyConstraints: [MetaschemaPrimaryKeyConstraint]
          uniqueConstraints: [MetaschemaUniqueConstraint]
          checkConstraints: [MetaschemaCheckConstraint]
          exclusionConstraints: [MetaschemaExclusionConstraint]
        }
        union MetaschemaConstraint =
            MetaschemaForeignKeyConstraint
          | MetaschemaUniqueConstraint
          | MetaschemaPrimaryKeyConstraint
          | MetaschemaCheckConstraint
          | MetaschemaExclusionConstraint
        type MetaschemaForeignKeyConstraint {
          name: String!
          fields: [MetaschemaField]
          refTable: MetaschemaTable
          refFields: [MetaschemaField]
        }
        type MetaschemaUniqueConstraint {
          name: String!
          fields: [MetaschemaField]
        }
        type MetaschemaPrimaryKeyConstraint {
          name: String!
          fields: [MetaschemaField]
        }
        type MetaschemaCheckConstraint {
          name: String!
          fields: [MetaschemaField]
        }
        type MetaschemaExclusionConstraint {
          name: String!
          fields: [MetaschemaField]
        }
        type Metaschema {
          tables: [MetaschemaTable]
        }
        extend type Query {
          _meta: Metaschema
        }
      `,
      resolvers: {
        // TODO determine why check constraints aren't coming through
        MetaschemaCheckConstraint: {
          /** @param constraint {import('graphile-build-pg').PgConstraint} */
          fields(constraint) {
            return constraint.keyAttributes;
          }
        },
        MetaschemaExclusionConstraint: {
          /** @param constraint {import('graphile-build-pg').PgConstraint} */
          fields(constraint) {
            return constraint.keyAttributes;
          }
        },
        MetaschemaUniqueConstraint: {
          /** @param constraint {import('graphile-build-pg').PgConstraint} */
          fields(constraint) {
            return constraint.keyAttributes;
          }
        },
        MetaschemaPrimaryKeyConstraint: {
          /** @param constraint {import('graphile-build-pg').PgConstraint} */
          fields(constraint) {
            return constraint.keyAttributes;
          }
        },
        MetaschemaForeignKeyConstraint: {
          /** @param constraint {import('graphile-build-pg').PgConstraint} */
          fields(constraint) {
            return constraint.keyAttributes;
          },
          /** @param constraint {import('graphile-build-pg').PgConstraint} */
          refTable(constraint) {
            return constraint.foreignClass;
          },
          /** @param constraint {import('graphile-build-pg').PgConstraint} */
          refFields(constraint) {
            return constraint.foreignKeyAttributes;
          }
        },
        MetaschemaType: {
          /** @param attr {import('graphile-build-pg').PgType} */
          name(type) {
            // TODO what is the best API here?
            // 1. we could return original _name, e.g. _citext (= citext[])
            // 2. we could return original type name and include isArray
            if (type.isPgArray && type.arrayItemType?.name) {
              return type.arrayItemType.name;
            }
            return type.name;
          },
          typmod(type) {
            const modifier = type.attrTypeModifier;
            if (!modifier) return null;

            if (type.name === 'geography' || type.name === 'geometry') {
              // Ref: https://github.com/postgis/postgis/blob/2.5.2/liblwgeom/liblwgeom.h.in#L156-L173
              // #define TYPMOD_GET_SRID(typmod) ((((typmod) & 0x0FFFFF00) - ((typmod) & 0x10000000)) >> 8)
              // #define TYPMOD_GET_TYPE(typmod) ((typmod & 0x000000FC)>>2)
              // #define TYPMOD_GET_Z(typmod) ((typmod & 0x00000002)>>1)
              // #define TYPMOD_GET_M(typmod) (typmod & 0x00000001)
              const srid =
                ((modifier & 0x0fffff00) - (modifier & 0x10000000)) >> 8;
              const subtype = (modifier & 0x000000fc) >> 2;
              const hasZ = (modifier & 0x00000002) >> 1 === 1;
              const hasM = (modifier & 0x00000001) === 1;
              if (subtype < GIS_TYPES.length) {
                return {
                  srid,
                  subtype,
                  hasZ,
                  hasM,
                  gisType: GIS_TYPES[subtype]
                };
              }
            }
            return { modifier };
          },
          modifier(type) {
            return type.attrTypeModifier;
          },
          isArray(type) {
            return type.isPgArray;
          }
        },
        MetaschemaField: {
          /** @param attr {import('graphile-build-pg').PgAttribute} */
          name(attr) {
            return inflection.column(attr);
          },
          /** @param attr {import('graphile-build-pg').PgAttribute} */
          type(attr) {
            if (attr.typeModifier > 0) {
              return {
                ...attr.type,
                attrTypeModifier: attr.typeModifier
              };
            }
            return attr.type;
          }
        },
        MetaschemaTableInflection: {
          deleteByPrimaryKey(table) {
            if (!table.primaryKeyConstraint?.keyAttributes?.length) return null;
            return inflection.deleteByKeys(
              table.primaryKeyConstraint.keyAttributes,
              table,
              table.primaryKeyConstraint
            );
          },
          updateByPrimaryKey(table) {
            if (!table.primaryKeyConstraint?.keyAttributes?.length) return null;
            return inflection.updateByKeys(
              table.primaryKeyConstraint.keyAttributes,
              table,
              table.primaryKeyConstraint
            );
          },
          createField(table) {
            return inflection.createField(table);
          },
          allRows(table) {
            return inflection.allRows(table);
          },
          allRowsSimple(table) {
            return inflection.allRowsSimple(table);
          },
          tableFieldName(table) {
            return inflection.tableFieldName(table);
          },
          tableType(table) {
            return inflection.tableType(table);
          },
          createPayloadType(table) {
            return inflection.createPayloadType(table);
          },
          updatePayloadType(table) {
            return inflection.updatePayloadType(table);
          },
          deletePayloadType(table) {
            return inflection.deletePayloadType(table);
          }
        },
        MetaschemaTable: {
          /** @param table {import('graphile-build-pg').PgClass} */
          name(table) {
            return inflection.tableType(table);
            // return inflection._tableName(table);
          },
          /** @param table {import('graphile-build-pg').PgClass} */
          fields(table) {
            return table.attributes.filter(attr => {
              if (attr.num < 1) return false; // low-level props
              return true;
            });
          },
          /** @param table {import('graphile-build-pg').PgClass} */
          inflection(table) {
            // return table so the MetaschemaTableInflection resolver uses that as input
            return table;
          },
          /** @param table {import('graphile-build-pg').PgClass} */
          constraints(table) {
            return table.constraints;
          },
          /** @param table {import('graphile-build-pg').PgClass} */
          foreignKeyConstraints(table) {
            return table.constraints.filter(c => c.type === 'f');
          },
          /** @param table {import('graphile-build-pg').PgClass} */
          primaryKeyConstraints(table) {
            return table.constraints.filter(c => c.type === 'p');
          },
          /** @param table {import('graphile-build-pg').PgClass} */
          uniqueConstraints(table) {
            return table.constraints.filter(c => c.type === 'u');
          },
          /** @param table {import('graphile-build-pg').PgClass} */
          checkConstraints(table) {
            return table.constraints.filter(c => c.type === 'c');
          },
          /** @param table {import('graphile-build-pg').PgClass} */
          exclusionConstraints(table) {
            return table.constraints.filter(c => c.type === 'x');
          }
        },
        MetaschemaConstraint: {
          /** @param obj {import('graphile-build-pg').PgConstraint} */
          __resolveType(obj) {
            switch (obj.type) {
              case 'p':
                return 'MetaschemaPrimaryKeyConstraint';
              case 'f':
                return 'MetaschemaForeignKeyConstraint';
              case 'c':
                return 'MetaschemaCheckConstraint';
              case 'u':
                return 'MetaschemaUniqueConstraint';
              case 'x':
                return 'MetaschemaExclusionConstraint';
            }
          }
        },
        Metaschema: {
          tables() {
            return introspection.class.filter(kls => {
              if (!schemas.includes(kls.namespaceName)) return false;
              // r = ordinary table, i = index, S = sequence, t = TOAST table, v = view, m = materialized view, c = composite type, f = foreign table, p = partitioned table, I = partitioned index
              if (kls.classKind !== 'r') return false;
              return true;
            });
          }
        },
        Query: {
          _meta() {
            // just placeholder
            return {};
          }
        }
      }
    };
  }
);

export default PgMetaschemaPlugin;
