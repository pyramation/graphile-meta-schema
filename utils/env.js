process.env.SCHEMA = 'app_meta';
process.env.PGDATABASE = 'metaschema_example';
