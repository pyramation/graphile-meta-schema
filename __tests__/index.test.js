import '../utils/env';
import { GraphQLTest, env, snapshot } from 'graphile-test';
import { GetMetaSchema, GetMetaSchemaUnion } from '../utils/queries';
import { PgMetaschemaPlugin } from '../src';

const { SCHEMA } = env;

const getDbString = () =>
  `postgres://${env.PGUSER}:${env.PGPASSWORD}@${env.PGHOST}:${env.PGPORT}/${env.PGDATABASE}`;

const { setup, teardown, graphQL } = GraphQLTest(
  {
    appendPlugins: [PgMetaschemaPlugin],
    dynamicJson: true,
    schema: SCHEMA,
    graphqlRoute: '/graphql'
  },
  getDbString()
);

beforeAll(async () => {
  await setup();
});
afterAll(async () => {
  await teardown();
});

it('individual', async () => {
  await graphQL(async query => {
    const data = await query(GetMetaSchema);
    expect(snapshot(data)).toMatchSnapshot();
  });
});
it('union', async () => {
  await graphQL(async query => {
    const data = await query(GetMetaSchemaUnion);
    expect(snapshot(data)).toMatchSnapshot();
  });
});
