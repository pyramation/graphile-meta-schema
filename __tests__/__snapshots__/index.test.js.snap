// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`individual 1`] = `
Object {
  "data": Object {
    "_meta": Object {
      "tables": Array [
        Object {
          "foreignKeyConstraints": Array [],
          "name": "BoundingBox",
          "primaryKeyConstraints": Array [
            Object {
              "name": "bounding_box_pkey",
            },
          ],
          "uniqueConstraints": Array [],
        },
        Object {
          "foreignKeyConstraints": Array [
            Object {
              "name": "permission_assignment_perm_id_fkey",
            },
            Object {
              "name": "permission_assignment_role_id_fkey",
            },
          ],
          "name": "PermissionAssignment",
          "primaryKeyConstraints": Array [
            Object {
              "name": "permission_assignment_pkey",
            },
          ],
          "uniqueConstraints": Array [],
        },
        Object {
          "foreignKeyConstraints": Array [],
          "name": "Permission",
          "primaryKeyConstraints": Array [
            Object {
              "name": "permissions_pkey",
            },
          ],
          "uniqueConstraints": Array [],
        },
        Object {
          "foreignKeyConstraints": Array [
            Object {
              "name": "roles_org_id_fkey",
            },
          ],
          "name": "Role",
          "primaryKeyConstraints": Array [
            Object {
              "name": "roles_pkey",
            },
          ],
          "uniqueConstraints": Array [],
        },
        Object {
          "foreignKeyConstraints": Array [
            Object {
              "name": "subject_assignment_subj_id_fkey",
            },
            Object {
              "name": "subject_assignment_role_id_fkey",
            },
          ],
          "name": "SubjectAssignment",
          "primaryKeyConstraints": Array [
            Object {
              "name": "subject_assignment_pkey",
            },
          ],
          "uniqueConstraints": Array [],
        },
        Object {
          "foreignKeyConstraints": Array [
            Object {
              "name": "user_settings_user_id_fkey",
            },
          ],
          "name": "UserSetting",
          "primaryKeyConstraints": Array [
            Object {
              "name": "user_settings_pkey",
            },
          ],
          "uniqueConstraints": Array [],
        },
        Object {
          "foreignKeyConstraints": Array [],
          "name": "User",
          "primaryKeyConstraints": Array [
            Object {
              "name": "users_pkey",
            },
          ],
          "uniqueConstraints": Array [
            Object {
              "name": "users_username_key",
            },
          ],
        },
        Object {
          "foreignKeyConstraints": Array [],
          "name": "Postcode",
          "primaryKeyConstraints": Array [
            Object {
              "name": "zip_codes_pkey",
            },
          ],
          "uniqueConstraints": Array [],
        },
      ],
    },
  },
}
`;

exports[`union 1`] = `
Object {
  "data": Object {
    "_meta": Object {
      "tables": Array [
        Object {
          "constraints": Array [
            Object {
              "__typename": "MetaschemaPrimaryKeyConstraint",
              "fields": Array [
                Object {
                  "name": "id",
                  "type": Object {
                    "modifier": null,
                    "name": "Int",
                    "pg": "int4",
                    "subtype": null,
                    "typmod": null,
                  },
                },
              ],
              "name": "bounding_box_pkey",
            },
          ],
          "fields": Array [
            Object {
              "name": "id",
              "type": Object {
                "modifier": null,
                "name": "Int",
                "pg": "int4",
                "subtype": null,
                "typmod": null,
              },
            },
            Object {
              "name": "zip",
              "type": Object {
                "modifier": null,
                "name": "Int",
                "pg": "int4",
                "subtype": null,
                "typmod": null,
              },
            },
            Object {
              "name": "location",
              "type": Object {
                "modifier": 1107460,
                "name": "String",
                "pg": "geometry",
                "subtype": null,
                "typmod": Object {
                  "gisType": "Point",
                  "hasM": false,
                  "hasZ": false,
                  "srid": 4326,
                  "subtype": 1,
                },
              },
            },
            Object {
              "name": "bbox",
              "type": Object {
                "modifier": 1107468,
                "name": "String",
                "pg": "geometry",
                "subtype": null,
                "typmod": Object {
                  "gisType": "Polygon",
                  "hasM": false,
                  "hasZ": false,
                  "srid": 4326,
                  "subtype": 3,
                },
              },
            },
          ],
          "inflection": Object {
            "allRows": "allBoundingBoxes",
            "allRowsSimple": "allBoundingBoxesList",
            "conditionType": "BoundingBoxCondition",
            "connection": "BoundingBoxesConnection",
            "createField": "createBoundingBox",
            "createInputType": "CreateBoundingBoxInput",
            "createPayloadType": "CreateBoundingBoxPayload",
            "deleteByPrimaryKey": "deleteBoundingBoxById",
            "deletePayloadType": "DeleteBoundingBoxPayload",
            "edge": "BoundingBoxesEdge",
            "edgeField": "boundingBoxEdge",
            "enumType": "BoundingBox",
            "inputType": "BoundingBoxInput",
            "orderByType": "BoundingBoxesOrderBy",
            "patchField": "boundingBoxPatch",
            "patchType": "BoundingBoxPatch",
            "tableFieldName": "boundingBox",
            "tableType": "BoundingBox",
            "typeName": "bounding_box",
            "updateByPrimaryKey": "updateBoundingBoxById",
            "updatePayloadType": "UpdateBoundingBoxPayload",
          },
          "name": "BoundingBox",
          "query": Object {
            "all": "allBoundingBoxes",
            "create": "createBoundingBox",
            "delete": "deleteBoundingBoxById",
            "one": "boundingBox",
            "update": "updateBoundingBoxById",
          },
        },
        Object {
          "constraints": Array [
            Object {
              "fields": Array [
                Object {
                  "name": "permId",
                },
              ],
              "refFields": Array [
                Object {
                  "name": "id",
                },
              ],
              "refTable": Object {
                "fields": Array [
                  Object {
                    "name": "id",
                  },
                  Object {
                    "name": "name",
                  },
                ],
                "name": "Permission",
              },
            },
            Object {
              "__typename": "MetaschemaPrimaryKeyConstraint",
              "fields": Array [
                Object {
                  "name": "permId",
                  "type": Object {
                    "modifier": null,
                    "name": "BigInt",
                    "pg": "int8",
                    "subtype": null,
                    "typmod": null,
                  },
                },
                Object {
                  "name": "roleId",
                  "type": Object {
                    "modifier": null,
                    "name": "BigInt",
                    "pg": "int8",
                    "subtype": null,
                    "typmod": null,
                  },
                },
              ],
              "name": "permission_assignment_pkey",
            },
            Object {
              "fields": Array [
                Object {
                  "name": "roleId",
                },
              ],
              "refFields": Array [
                Object {
                  "name": "id",
                },
              ],
              "refTable": Object {
                "fields": Array [
                  Object {
                    "name": "id",
                  },
                  Object {
                    "name": "orgId",
                  },
                ],
                "name": "Role",
              },
            },
          ],
          "fields": Array [
            Object {
              "name": "permId",
              "type": Object {
                "modifier": null,
                "name": "BigInt",
                "pg": "int8",
                "subtype": null,
                "typmod": null,
              },
            },
            Object {
              "name": "roleId",
              "type": Object {
                "modifier": null,
                "name": "BigInt",
                "pg": "int8",
                "subtype": null,
                "typmod": null,
              },
            },
          ],
          "inflection": Object {
            "allRows": "allPermissionAssignments",
            "allRowsSimple": "allPermissionAssignmentsList",
            "conditionType": "PermissionAssignmentCondition",
            "connection": "PermissionAssignmentsConnection",
            "createField": "createPermissionAssignment",
            "createInputType": "CreatePermissionAssignmentInput",
            "createPayloadType": "CreatePermissionAssignmentPayload",
            "deleteByPrimaryKey": "deletePermissionAssignmentByPermIdAndRoleId",
            "deletePayloadType": "DeletePermissionAssignmentPayload",
            "edge": "PermissionAssignmentsEdge",
            "edgeField": "permissionAssignmentEdge",
            "enumType": "PermissionAssignment",
            "inputType": "PermissionAssignmentInput",
            "orderByType": "PermissionAssignmentsOrderBy",
            "patchField": "permissionAssignmentPatch",
            "patchType": "PermissionAssignmentPatch",
            "tableFieldName": "permissionAssignment",
            "tableType": "PermissionAssignment",
            "typeName": "permission_assignment",
            "updateByPrimaryKey": "updatePermissionAssignmentByPermIdAndRoleId",
            "updatePayloadType": "UpdatePermissionAssignmentPayload",
          },
          "name": "PermissionAssignment",
          "query": Object {
            "all": "allPermissionAssignments",
            "create": "createPermissionAssignment",
            "delete": "deletePermissionAssignmentByPermIdAndRoleId",
            "one": "permissionAssignment",
            "update": "updatePermissionAssignmentByPermIdAndRoleId",
          },
        },
        Object {
          "constraints": Array [
            Object {
              "__typename": "MetaschemaPrimaryKeyConstraint",
              "fields": Array [
                Object {
                  "name": "id",
                  "type": Object {
                    "modifier": null,
                    "name": "Int",
                    "pg": "int4",
                    "subtype": null,
                    "typmod": null,
                  },
                },
              ],
              "name": "permissions_pkey",
            },
          ],
          "fields": Array [
            Object {
              "name": "id",
              "type": Object {
                "modifier": null,
                "name": "Int",
                "pg": "int4",
                "subtype": null,
                "typmod": null,
              },
            },
            Object {
              "name": "name",
              "type": Object {
                "modifier": null,
                "name": "String",
                "pg": "citext",
                "subtype": null,
                "typmod": null,
              },
            },
          ],
          "inflection": Object {
            "allRows": "allPermissions",
            "allRowsSimple": "allPermissionsList",
            "conditionType": "PermissionCondition",
            "connection": "PermissionsConnection",
            "createField": "createPermission",
            "createInputType": "CreatePermissionInput",
            "createPayloadType": "CreatePermissionPayload",
            "deleteByPrimaryKey": "deletePermissionById",
            "deletePayloadType": "DeletePermissionPayload",
            "edge": "PermissionsEdge",
            "edgeField": "permissionEdge",
            "enumType": "Permissions",
            "inputType": "PermissionInput",
            "orderByType": "PermissionsOrderBy",
            "patchField": "permissionPatch",
            "patchType": "PermissionPatch",
            "tableFieldName": "permission",
            "tableType": "Permission",
            "typeName": "permissions",
            "updateByPrimaryKey": "updatePermissionById",
            "updatePayloadType": "UpdatePermissionPayload",
          },
          "name": "Permission",
          "query": Object {
            "all": "allPermissions",
            "create": "createPermission",
            "delete": "deletePermissionById",
            "one": "permission",
            "update": "updatePermissionById",
          },
        },
        Object {
          "constraints": Array [
            Object {
              "__typename": "MetaschemaPrimaryKeyConstraint",
              "fields": Array [
                Object {
                  "name": "id",
                  "type": Object {
                    "modifier": null,
                    "name": "Int",
                    "pg": "int4",
                    "subtype": null,
                    "typmod": null,
                  },
                },
              ],
              "name": "roles_pkey",
            },
            Object {
              "fields": Array [
                Object {
                  "name": "orgId",
                },
              ],
              "refFields": Array [
                Object {
                  "name": "id",
                },
              ],
              "refTable": Object {
                "fields": Array [
                  Object {
                    "name": "id",
                  },
                  Object {
                    "name": "username",
                  },
                ],
                "name": "User",
              },
            },
          ],
          "fields": Array [
            Object {
              "name": "id",
              "type": Object {
                "modifier": null,
                "name": "Int",
                "pg": "int4",
                "subtype": null,
                "typmod": null,
              },
            },
            Object {
              "name": "orgId",
              "type": Object {
                "modifier": null,
                "name": "BigInt",
                "pg": "int8",
                "subtype": null,
                "typmod": null,
              },
            },
          ],
          "inflection": Object {
            "allRows": "allRoles",
            "allRowsSimple": "allRolesList",
            "conditionType": "RoleCondition",
            "connection": "RolesConnection",
            "createField": "createRole",
            "createInputType": "CreateRoleInput",
            "createPayloadType": "CreateRolePayload",
            "deleteByPrimaryKey": "deleteRoleById",
            "deletePayloadType": "DeleteRolePayload",
            "edge": "RolesEdge",
            "edgeField": "roleEdge",
            "enumType": "Roles",
            "inputType": "RoleInput",
            "orderByType": "RolesOrderBy",
            "patchField": "rolePatch",
            "patchType": "RolePatch",
            "tableFieldName": "role",
            "tableType": "Role",
            "typeName": "roles",
            "updateByPrimaryKey": "updateRoleById",
            "updatePayloadType": "UpdateRolePayload",
          },
          "name": "Role",
          "query": Object {
            "all": "allRoles",
            "create": "createRole",
            "delete": "deleteRoleById",
            "one": "role",
            "update": "updateRoleById",
          },
        },
        Object {
          "constraints": Array [
            Object {
              "fields": Array [
                Object {
                  "name": "subjId",
                },
              ],
              "refFields": Array [
                Object {
                  "name": "id",
                },
              ],
              "refTable": Object {
                "fields": Array [
                  Object {
                    "name": "id",
                  },
                  Object {
                    "name": "username",
                  },
                ],
                "name": "User",
              },
            },
            Object {
              "__typename": "MetaschemaPrimaryKeyConstraint",
              "fields": Array [
                Object {
                  "name": "subjId",
                  "type": Object {
                    "modifier": null,
                    "name": "BigInt",
                    "pg": "int8",
                    "subtype": null,
                    "typmod": null,
                  },
                },
                Object {
                  "name": "roleId",
                  "type": Object {
                    "modifier": null,
                    "name": "BigInt",
                    "pg": "int8",
                    "subtype": null,
                    "typmod": null,
                  },
                },
              ],
              "name": "subject_assignment_pkey",
            },
            Object {
              "fields": Array [
                Object {
                  "name": "roleId",
                },
              ],
              "refFields": Array [
                Object {
                  "name": "id",
                },
              ],
              "refTable": Object {
                "fields": Array [
                  Object {
                    "name": "id",
                  },
                  Object {
                    "name": "orgId",
                  },
                ],
                "name": "Role",
              },
            },
          ],
          "fields": Array [
            Object {
              "name": "subjId",
              "type": Object {
                "modifier": null,
                "name": "BigInt",
                "pg": "int8",
                "subtype": null,
                "typmod": null,
              },
            },
            Object {
              "name": "roleId",
              "type": Object {
                "modifier": null,
                "name": "BigInt",
                "pg": "int8",
                "subtype": null,
                "typmod": null,
              },
            },
          ],
          "inflection": Object {
            "allRows": "allSubjectAssignments",
            "allRowsSimple": "allSubjectAssignmentsList",
            "conditionType": "SubjectAssignmentCondition",
            "connection": "SubjectAssignmentsConnection",
            "createField": "createSubjectAssignment",
            "createInputType": "CreateSubjectAssignmentInput",
            "createPayloadType": "CreateSubjectAssignmentPayload",
            "deleteByPrimaryKey": "deleteSubjectAssignmentBySubjIdAndRoleId",
            "deletePayloadType": "DeleteSubjectAssignmentPayload",
            "edge": "SubjectAssignmentsEdge",
            "edgeField": "subjectAssignmentEdge",
            "enumType": "SubjectAssignment",
            "inputType": "SubjectAssignmentInput",
            "orderByType": "SubjectAssignmentsOrderBy",
            "patchField": "subjectAssignmentPatch",
            "patchType": "SubjectAssignmentPatch",
            "tableFieldName": "subjectAssignment",
            "tableType": "SubjectAssignment",
            "typeName": "subject_assignment",
            "updateByPrimaryKey": "updateSubjectAssignmentBySubjIdAndRoleId",
            "updatePayloadType": "UpdateSubjectAssignmentPayload",
          },
          "name": "SubjectAssignment",
          "query": Object {
            "all": "allSubjectAssignments",
            "create": "createSubjectAssignment",
            "delete": "deleteSubjectAssignmentBySubjIdAndRoleId",
            "one": "subjectAssignment",
            "update": "updateSubjectAssignmentBySubjIdAndRoleId",
          },
        },
        Object {
          "constraints": Array [
            Object {
              "__typename": "MetaschemaPrimaryKeyConstraint",
              "fields": Array [
                Object {
                  "name": "userId",
                  "type": Object {
                    "modifier": null,
                    "name": "BigInt",
                    "pg": "int8",
                    "subtype": null,
                    "typmod": null,
                  },
                },
              ],
              "name": "user_settings_pkey",
            },
            Object {
              "fields": Array [
                Object {
                  "name": "userId",
                },
              ],
              "refFields": Array [
                Object {
                  "name": "id",
                },
              ],
              "refTable": Object {
                "fields": Array [
                  Object {
                    "name": "id",
                  },
                  Object {
                    "name": "username",
                  },
                ],
                "name": "User",
              },
            },
          ],
          "fields": Array [
            Object {
              "name": "userId",
              "type": Object {
                "modifier": null,
                "name": "BigInt",
                "pg": "int8",
                "subtype": null,
                "typmod": null,
              },
            },
            Object {
              "name": "setting1",
              "type": Object {
                "modifier": null,
                "name": "String",
                "pg": "text",
                "subtype": null,
                "typmod": null,
              },
            },
          ],
          "inflection": Object {
            "allRows": "allUserSettings",
            "allRowsSimple": "allUserSettingsList",
            "conditionType": "UserSettingCondition",
            "connection": "UserSettingsConnection",
            "createField": "createUserSetting",
            "createInputType": "CreateUserSettingInput",
            "createPayloadType": "CreateUserSettingPayload",
            "deleteByPrimaryKey": "deleteUserSettingByUserId",
            "deletePayloadType": "DeleteUserSettingPayload",
            "edge": "UserSettingsEdge",
            "edgeField": "userSettingEdge",
            "enumType": "UserSettings",
            "inputType": "UserSettingInput",
            "orderByType": "UserSettingsOrderBy",
            "patchField": "userSettingPatch",
            "patchType": "UserSettingPatch",
            "tableFieldName": "userSetting",
            "tableType": "UserSetting",
            "typeName": "user_settings",
            "updateByPrimaryKey": "updateUserSettingByUserId",
            "updatePayloadType": "UpdateUserSettingPayload",
          },
          "name": "UserSetting",
          "query": Object {
            "all": "allUserSettings",
            "create": "createUserSetting",
            "delete": "deleteUserSettingByUserId",
            "one": "userSetting",
            "update": "updateUserSettingByUserId",
          },
        },
        Object {
          "constraints": Array [
            Object {
              "__typename": "MetaschemaPrimaryKeyConstraint",
              "fields": Array [
                Object {
                  "name": "id",
                  "type": Object {
                    "modifier": null,
                    "name": "Int",
                    "pg": "int4",
                    "subtype": null,
                    "typmod": null,
                  },
                },
              ],
              "name": "users_pkey",
            },
            Object {
              "__typename": "MetaschemaUniqueConstraint",
              "fields": Array [
                Object {
                  "name": "username",
                  "type": Object {
                    "modifier": null,
                    "name": "String",
                    "pg": "citext",
                    "subtype": null,
                    "typmod": null,
                  },
                },
              ],
              "name": "users_username_key",
            },
          ],
          "fields": Array [
            Object {
              "name": "id",
              "type": Object {
                "modifier": null,
                "name": "Int",
                "pg": "int4",
                "subtype": null,
                "typmod": null,
              },
            },
            Object {
              "name": "username",
              "type": Object {
                "modifier": null,
                "name": "String",
                "pg": "citext",
                "subtype": null,
                "typmod": null,
              },
            },
          ],
          "inflection": Object {
            "allRows": "allUsers",
            "allRowsSimple": "allUsersList",
            "conditionType": "UserCondition",
            "connection": "UsersConnection",
            "createField": "createUser",
            "createInputType": "CreateUserInput",
            "createPayloadType": "CreateUserPayload",
            "deleteByPrimaryKey": "deleteUserById",
            "deletePayloadType": "DeleteUserPayload",
            "edge": "UsersEdge",
            "edgeField": "userEdge",
            "enumType": "Users",
            "inputType": "UserInput",
            "orderByType": "UsersOrderBy",
            "patchField": "userPatch",
            "patchType": "UserPatch",
            "tableFieldName": "user",
            "tableType": "User",
            "typeName": "users",
            "updateByPrimaryKey": "updateUserById",
            "updatePayloadType": "UpdateUserPayload",
          },
          "name": "User",
          "query": Object {
            "all": "allUsers",
            "create": "createUser",
            "delete": "deleteUserById",
            "one": "user",
            "update": "updateUserById",
          },
        },
        Object {
          "constraints": Array [
            Object {
              "__typename": "MetaschemaPrimaryKeyConstraint",
              "fields": Array [
                Object {
                  "name": "id",
                  "type": Object {
                    "modifier": null,
                    "name": "Int",
                    "pg": "int4",
                    "subtype": null,
                    "typmod": null,
                  },
                },
              ],
              "name": "zip_codes_pkey",
            },
          ],
          "fields": Array [
            Object {
              "name": "id",
              "type": Object {
                "modifier": null,
                "name": "Int",
                "pg": "int4",
                "subtype": null,
                "typmod": null,
              },
            },
            Object {
              "name": "zipCode",
              "type": Object {
                "modifier": null,
                "name": "Int",
                "pg": "int4",
                "subtype": null,
                "typmod": null,
              },
            },
            Object {
              "name": "location",
              "type": Object {
                "modifier": 1107460,
                "name": "String",
                "pg": "geometry",
                "subtype": null,
                "typmod": Object {
                  "gisType": "Point",
                  "hasM": false,
                  "hasZ": false,
                  "srid": 4326,
                  "subtype": 1,
                },
              },
            },
          ],
          "inflection": Object {
            "allRows": "allPostcodes",
            "allRowsSimple": "allPostcodesList",
            "conditionType": "PostcodeCondition",
            "connection": "PostcodesConnection",
            "createField": "createPostcode",
            "createInputType": "CreatePostcodeInput",
            "createPayloadType": "CreatePostcodePayload",
            "deleteByPrimaryKey": "deletePostcodeById",
            "deletePayloadType": "DeletePostcodePayload",
            "edge": "PostcodesEdge",
            "edgeField": "postcodeEdge",
            "enumType": "Postcode",
            "inputType": "PostcodeInput",
            "orderByType": "PostcodesOrderBy",
            "patchField": "postcodePatch",
            "patchType": "PostcodePatch",
            "tableFieldName": "postcode",
            "tableType": "Postcode",
            "typeName": "postcode",
            "updateByPrimaryKey": "updatePostcodeById",
            "updatePayloadType": "UpdatePostcodePayload",
          },
          "name": "Postcode",
          "query": Object {
            "all": "allPostcodes",
            "create": "createPostcode",
            "delete": "deletePostcodeById",
            "one": "postcode",
            "update": "updatePostcodeById",
          },
        },
      ],
    },
  },
}
`;
